arctic/store/_version_store_utils.py

```diff
@@ -76,6 +80,219 @@ def _define_compat_pickle_load():
+                'PREV_MISSING' if prev_n < n - 1 else '',
+                'CORRUPTED VERSION' if corrupted else '')
+            )
+        prev_rows = v.get('up_to', 0)
+        prev_n = n
+        prev_v = v
+
+    logging.info('\nSegments for {}:'.format(sym))
+    for seg in l._collection.find({'symbol': sym}, sort=[('_id', pymongo.ASCENDING)]):
+        logging.info("{: <32}  {: <7}  {: <10} {: <30}".format(
+            hashlib.md5(seg['sha']).hexdigest(),
+            seg.get('segment'),
+            'compressed' if seg.get('compressed', False) else 'raw',
+            [str(p) for p in seg.get('parent', [])]
+        ))
+
+
+def _fast_check_corruption(collection, sym, v, check_count=True, check_last_segment=True, check_append_safe=False):
+    if v is None:
+        logging.warning("Symbol {} with version {} not found, so can't be corrupted.".format(sym, v))
+        return False
+    
+    if not check_count and not check_last_segment:
+        raise ValueError("_fast_check_corruption must be called with either of "
+                         "check_count and check_last_segment set to True")
+
+    # If version marked symbol as deleted, it will force writes/appends to start from a new base: non corrupted.
+    if isinstance(v.get('metadata'), dict) and v['metadata'].get('deleted'):
+        return False
+     
+    if check_append_safe:
+        # Check whether appending to the symbol version can potentially corrupt the data (history branch).
+        # Inspect all segments, don't limit to v['up_to']. No newer append segments after v should exist.
+        spec = {'symbol': sym, 'parent': v.get('base_version_id', v['_id'])}
+    else:
+        # Only verify segment count for current symbol version, don't check corruptability of future appends.
+        spec = {'symbol': sym, 'parent': v.get('base_version_id', v['_id']), 'segment': {'$lt': v['up_to']}}
+
+    try:
+        # Not that commands sequence (a) is slower than (b)
+        # (a) curs = collection.find(spec, {'segment': 1}, sort=[('segment', pymongo.DESCENDING)])
+        #     curs.count()
+        #     curs.next()
+        # (b) collection.find(spec, {'segment': 1}).count()
+        #     collection.find_one(spec, {'segment': 1}, sort=[('segment', pymongo.DESCENDING)])
+
+        if check_count:
+            total_segments = collection.find(spec, {'segment': 1}).count()
+            # Quick check: compare segment count
+            if total_segments != v.get('segment_count', 0):
+                return True  # corrupted, don't proceed with fetching from mongo the first hit
+            # Quick check: Segment counts agree and size is zero
+            if total_segments == 0:
+                return False
+        
+        if check_last_segment:
+            # Quick check: compare the maximum segment's up_to number. It has to verify the version's up_to.
+            max_seg = collection.find_one(spec, {'segment': 1}, sort=[('segment', pymongo.DESCENDING)])
```

Comment:
Out of interest, what's the explain() on this look like?

Reply:
Look similar.
Without the projection:
https://gist.github.com/dimosped/b0eea9ce2b7084b2833adb4d63bacbac

With the projection:
https://gist.github.com/dimosped/70115b548a3daef71e9ceb5ef343187a


Also benchmarked and with projection was significantly faster, probably due to fethching more data:
200x rounds.

With projection Stats (seconds): 
	Min=0.000867128372192
	Max=0.0121870040894
	Mean=0.00173362374306
	Stdev=0.00144508307458

Without projection stats (seconds): 
	Min=0.00388884544373
	Max=0.0164110660553
	Mean=0.00772412419319
	Stdev=0.00237581682623